src/main/java/org/jenkinsci/plugins/workflow/libs/SCMBasedRetriever.java

```diff
@@ -0,0 +1,336 @@
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.libs;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Functions;
+import hudson.Util;
+import hudson.model.Computer;
+import hudson.model.Item;
+import hudson.model.Job;
+import hudson.model.Node;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.TopLevelItem;
+import hudson.model.listeners.ItemListener;
+import hudson.scm.SCM;
+import hudson.slaves.WorkspaceList;
+import hudson.util.FormValidation;
+import java.io.File;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.Callable;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.steps.scm.GenericSCMStep;
+import org.jenkinsci.plugins.workflow.steps.scm.SCMStep;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.verb.POST;
+
+/**
+ * Functionality common to {@link SCMSourceRetriever} and {@link SCMRetriever}.
+ */
+public abstract class SCMBasedRetriever extends LibraryRetriever {
+
+    private static final Logger LOGGER = Logger.getLogger(SCMBasedRetriever.class.getName());
+
+    @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Non-final for write access via the Script Console")
+    public static boolean INCLUDE_SRC_TEST_IN_LIBRARIES = Boolean.getBoolean(SCMSourceRetriever.class.getName() + ".INCLUDE_SRC_TEST_IN_LIBRARIES");
+
+    /**
+     * Matches ".." in positions where it would be treated as the parent directory.
+     *
+     * <p>Used to prevent {@link #libraryPath} from being used for directory traversal.
+     */
+    static final Pattern PROHIBITED_DOUBLE_DOT = Pattern.compile("(^|.*[\\\\/])\\.\\.($|[\\\\/].*)");
+
+    private boolean clone;
+
+    /**
+     * The path to the library inside of the SCM.
+     *
+     * {@code null} is the default and means that the library is in the root of the repository. Otherwise, the value is
+     * considered to be a relative path inside of the repository and always ends in a forward slash
+     *
+     * @see #setLibraryPath
+     */
+    private @CheckForNull String libraryPath;
+
+    public boolean isClone() {
+        return clone;
+    }
+
+    @DataBoundSetter public void setClone(boolean clone) {
+        this.clone = clone;
+    }
+
+    public String getLibraryPath() {
+        return libraryPath;
```

Comment:
NiT: Can this eventually be part of an URL at some point of the usages? Like an API call somewhere in the execution stack?

Reply:
Not sure what you are asking. This is a getter matching `setLibraryPath` as required by databinding (so you can round-trip a form in which a library is already defined).